### PR TITLE
Update jclouds, picking up Guice from core

### PR DESCRIPTION
**Pull request by jglick**
Derived from #189. Follows up #244 to pick up https://github.com/jenkinsci/jenkins/pull/5858. https://github.com/jenkinsci/artifact-manager-s3-plugin/pull/226#issuecomment-972041078 tracks a further jclouds upgrade. Possibly solves [JENKINS-65824](https://issues.jenkins.io/browse/JENKINS-65824).

### CODE REVIEW

File: src/test/java/io/jenkins/plugins/artifact_manager_jclouds/MockApiMetadata.java [hunk 1]
```diff
@@ -222,6 +223,11 @@ public String putBlob(String containerName, Blob blob) throws IOException {
             return null;
         }
 
+        @Override
```
Comment (jglick) [inline src/test/java/io/jenkins/plugins/artifact_manager_jclouds/MockApiMetadata.java:226]: https://github.com/apache/jclouds/pull/94#discussion_r651256380

**Comment by jglick** [comment]
@grochoge assuming this build passes, can you please download the `*.hpi` which will then be linked from the _Incrementals_ check, install it via the **Advanced** tab of the plugin manager, and verify whether it indeed fixes [JENKINS-65824](https://issues.jenkins.io/browse/JENKINS-65824) for you?

File: pom.xml [hunk 2]
```diff
@@ -280,8 +280,8 @@
             <groupId>org.jenkins-ci.tools</groupId>
             <artifactId>maven-hpi-plugin</artifactId>
             <configuration>
-              <!-- JENKINS-50520: since we need a custom version of Guava -->
-              <maskClasses>com.google.common.</maskClasses>
+              <!-- JENKINS-50520: since we need a custom version of Guava and Guice -->
```
Comment (jtnord) [inline pom.xml:283]: you really don't want to go here - it does not work as demonstrated with jacoco. (it works in the plugin but depending on the Guice initialisation and what other plugins you have it will randomly blow up during initialisation for some users...)

Comment (jglick) [reply pom.xml:283]: Why do you think that? Whatever jclouds is using Guice for, it should be isolated from the stuff that Jenkins core and other plugins are using it for (extension loading).

What I would really like to do here is remove the jclouds dependency altogether, and rewrite the implementation to use the AWS SDK exclusively (it already has to use it to get presigned URL support for blobs, which jclouds does not support). The original idea had been to use jclouds as the multiplexing layer so that we could rapidly add support for lots of blob stores, but things never worked out that way (the Azure folks went with their own implementation), and it has been a source of frustration and opacity.

**Comment by jtnord** [comment]
> Why do you think that? Whatever jclouds is using Guice for, it should be isolated from the stuff that Jenkins core and other plugins are using it for (extension loading).

**It should be** but that was not what I saw in practice.  the Guice from this classloader can be used for other things and then you end up with 2 Jenkins' and a lot of mess...

For example Jacoco was using a plugin first classloader and bundling Guice and Guave.  Jacoco plugin worked fine, but when you had some other plugin that used Guice for injection (assuming it was coming from core) it blew up -> https://issues.jenkins.io/browse/JENKINS-65757 

The stack in https://issues.jenkins.io/browse/JENKINS-65757?focusedCommentId=409963&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-409963 was one examle of the complete explosion that doing this caused.   

Basically I think this needs testing with the plugins that blew up before with Jacoco.

**Comment by jglick** [comment]
> the Guice from this classloader can be used for other things

Sounds like a bug in core?

> testing with the plugins that blew up before with Jacoco

Is there a known way to reproduce? https://github.com/jenkinsci/depgraph-view-plugin/blob/5b67c31b2d234be79792c758526ff35a2bbdf1b0/src/main/java/hudson/plugins/depgraph_view/model/graph/DependencyGraphModule.java#L42-L46 looks suspicious.

**Comment by jglick** [comment]
`MinioIntegrationTest#artifactArchive` still passes with

```diff
diff --git pom.xml pom.xml
index 462064b..952e585 100644
--- pom.xml
+++ pom.xml
@@ -246,6 +246,12 @@
       <artifactId>guice</artifactId>
       <version>4.2.3</version>
     </dependency>
+    <dependency>
+      <groupId>org.jenkins-ci.plugins</groupId>
+      <artifactId>swarm</artifactId>
+      <version>3.27</version>
+      <scope>test</scope>
+    </dependency>
   </dependencies>
   <dependencyManagement>
     <dependencies>
```

Judging by https://github.com/search?q=user%3Ajenkinsci+user%3Acloudbees+%22extends+AbstractModule%22&type=Code this idiom is not uncommon.

**Comment by jglick** [comment]
I suspect https://github.com/jenkinsci/jenkins/blob/13a4a94095f11b4edb44d9ba1173aa33e856a54f/core/src/main/java/hudson/ExtensionFinder.java#L263 is the problem but I cannot figure out what concretely went astray.

File: pom.xml [hunk 3]
```diff
@@ -16,7 +16,7 @@
   <properties>
     <revision>1.18</revision>
     <changelist>-SNAPSHOT</changelist>
-    <jenkins.version>2.320</jenkins.version>
+    <jenkins.version>2.321</jenkins.version>
```
Comment (jglick) [inline pom.xml:19]: Note that the `bom` is still old. #242 tracks problems with that.

File: pom.xml [hunk 4]
```diff
@@ -265,9 +260,33 @@
...
   </dependencyManagement>
 
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-enforcer-plugin</artifactId>
+        <configuration>
+          <rules combine.children="append">
+            <requireSameVersions>
```
Comment (jtnord) [inline pom.xml:278]: Should this be in core if it is tightly coupled to guice version (at least until guice gets removed from core 🦄)

**Review by jtnord — APPROVED** [review]

**Review by basil — APPROVED** [review]

File: pom.xml (hunk 4 above)
Comment (jglick) [reply pom.xml:278]: > Should this be in core if it is tightly coupled to guice version

It could be, though I suppose core does not use it; or it could be added to the core BOM.

> at least until guice gets removed from core

Or I can remove jclouds from this plugin: https://github.com/jenkinsci/artifact-manager-s3-plugin/pull/215#discussion_r656197429

Comment (jtnord) [reply pom.xml:278]: > It could be, though I suppose core does not use it; or it could be added to the core BOM.

if it was just in the BOM then it can constantly drift (core gets a new Guice, this plugin potentially breaks).

> Or I can remove jclouds from this plugin: #215 (comment)

probably better.

Comment (jglick) [reply pom.xml:278]: > core gets a new Guice, this plugin potentially breaks

Potentially, though the test failure I saw was about core bundling Guice 5 and this plugin bundling Guice Assisted Inject 4; I hope breaking changes are limited to major versions which should not be too common. The Enforcer rule will also alert us to problems if are updating the core dep for other reasons. (_Might_ also work in PCT, not sure.)
